File: web/src/api/pipeline.ts
import { api } from './client'
import type { PipelineRun } from '../types'

export async function listPipelines(appId?: number): Promise<PipelineRun[]> {
  const r = await api.get<{ items: PipelineRun[] }>('/pipelines', {
    params: appId ? { app_id: appId } : undefined,
  })
  return r.data.items
}

export async function getPipeline(id: number): Promise<PipelineRun> {
  const r = await api.get<PipelineRun>(`/pipelines/${id}`)
  return r.data
}

// deployApp 触发部署。
//   - strategy='single'（默认）：顺序逐台、fail-fast
//   - strategy='rolling'：分批并行，batchSize 必传 >=1
//   - strategy='blue_green'：后端按 active_group 自动选目标组（active→opposite，空→blue）
//     需要 App 已配 nginx_host_id + nginx_upstream_name；前端不必传 target_group
export async function deployApp(
  appId: number,
  artifactId: number,
  strategy: 'single' | 'rolling' | 'blue_green' = 'single',
  batchSize?: number,
): Promise<PipelineRun> {
  const body: Record<string, unknown> = {
    artifact_id: artifactId,
    strategy,
  }
  if (strategy === 'rolling') {
    body.batch_size = batchSize ?? 1
  }
  const r = await api.post<PipelineRun>(`/apps/${appId}/deploy`, body)
  return r.data
}

// cancelPipeline 请求取消运行中的 pipeline。后端 202 异步生效；最终 status 由 execute 收口。
export async function cancelPipeline(id: number): Promise<void> {
  await api.post(`/pipelines/${id}/cancel`)
}

// rollbackApp 一键回滚：每个 deployment 退到自己的 previous_artifact_id。
// 没有任何 deployment 有 previous → 后端 400。
export async function rollbackApp(appId: number): Promise<PipelineRun> {
  const r = await api.post<PipelineRun>(`/apps/${appId}/rollback`)
  return r.data
}

File: internal/api/handler/pipeline.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"

	apperr "swift-devops/internal/pkg/errors"
	"swift-devops/internal/service"
)

// PipelineHandler 流水线触发与查询
type PipelineHandler struct {
	svc *service.PipelineService
}

func NewPipelineHandler(svc *service.PipelineService) *PipelineHandler {
	return &PipelineHandler{svc: svc}
}

// Deploy POST /apps/:id/deploy  body: {"artifact_id": N, "bundle_id": M, "strategy": "single|rolling", "batch_size": 2}
//
// Sprint X.6：body 加 bundle_id（多 service Bundle 链路，优先于 artifact_id）。
// 二选一：artifact_id 是旧单 jar 链路兼容；bundle_id 是新多 service 链路。
func (h *PipelineHandler) Deploy(c *gin.Context) {
	appID, ok := parseID(c)
	if !ok {
		return
	}
	var body struct {
		ArtifactID uint   `json:"artifact_id,omitempty"` // 旧链路；bundle_id==0 时必填
		BundleID   uint   `json:"bundle_id,omitempty"`   // Sprint X.6 新链路
		Strategy   string `json:"strategy,omitempty"`    // 默认 single
		BatchSize  int    `json:"batch_size,omitempty"`  // rolling 必填，>=1
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		apperr.Respond(c, apperr.Wrap(err, "BAD_REQUEST", err.Error(), http.StatusBadRequest))
		return
	}
	if body.ArtifactID == 0 && body.BundleID == 0 {
		apperr.Respond(c, apperr.New("BAD_REQUEST",
			"必须提供 artifact_id 或 bundle_id 之一", http.StatusBadRequest))
		return
	}
	actor := "unknown"
	if v, ok := c.Get("user"); ok {
		if s, ok := v.(string); ok {
			actor = s
		}
	}
	out, err := h.svc.Trigger(appID, body.ArtifactID, actor, service.TriggerOptions{
		Strategy:  body.Strategy,
		BatchSize: body.BatchSize,
		BundleID:  body.BundleID,
	})
	if err != nil {
		apperr.Respond(c, err)
		return
	}
	c.JSON(http.StatusAccepted, out)
}

// List GET /pipelines?app_id=N
func (h *PipelineHandler) List(c *gin.Context) {
	var appID uint
	if s := c.Query("app_id"); s != "" {
		n, err := strconv.ParseUint(s, 10, 32)
		if err != nil {
			apperr.Respond(c, apperr.New("BAD_REQUEST", "app_id 必须是整数", http.StatusBadRequest))
			return
		}
		appID = uint(n)
	}
	out, err := h.svc.List(appID)
	if err != nil {
		apperr.Respond(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"items": out})
}

// Get GET /pipelines/:id
func (h *PipelineHandler) Get(c *gin.Context) {
	id, ok := parseID(c)
	if !ok {
		return
	}
	out, err := h.svc.Get(id)
	if err != nil {
		apperr.Respond(c, err)
		return
	}
	c.JSON(http.StatusOK, out)
}

// Cancel POST /pipelines/:id/cancel
// 请求取消运行中的 pipeline。已结束的 run 返回 409。
// 取消是异步的：本接口立刻返回 202，执行线程会在最近的阶段间隙退出，
// 由 execute 收口把 status 落成 cancelled，并对未触达主机标 skipped。
func (h *PipelineHandler) Cancel(c *gin.Context) {
	id, ok := parseID(c)
	if !ok {
		return
	}
	if err := h.svc.Cancel(id); err != nil {
		apperr.Respond(c, err)
		return
	}
	c.JSON(http.StatusAccepted, gin.H{"id": id, "cancel": "requested"})
}

// Rollback POST /apps/:id/rollback
// 一键回滚：每个 deployment 退到自己的 previous_artifact_id。
// 没有任何 deployment 有 previous_artifact_id → 400 BAD_REQUEST。
func (h *PipelineHandler) Rollback(c *gin.Context) {
	appID, ok := parseID(c)
	if !ok {
		return
	}
	actor := "unknown"
	if v, ok := c.Get("user"); ok {
		if s, ok := v.(string); ok {
			actor = s
		}
	}
	out, err := h.svc.Rollback(appID, actor)
	if err != nil {
		apperr.Respond(c, err)
		return
	}
	c.JSON(http.StatusAccepted, out)
}
